src/00/02-introspect/proposal.adoc

```diff
@@ -0,0 +1,879 @@
+----
+
+=== Levels
+
+For some use cases it helps to simplify the extent of information the
+introspection understands. While it would be reasonable to expect a tool written
+in a modern general purpose programming language to fully implement all aspects
+of the introspection. It would not be practical to have a shell script parse
+and recognize the more challenging aspect of parsing version number ranges and
+matching them together. To support such use cases the introspection has to
+support levels "min" and "full".
+
+Obviously the "full" level equates to the tool understanding all the arguments
+and values. The "min" level only understands these:
+
+* Only introspection `--std-info` option.
+* Single version number in the responses for `--std-info`.
+
+This has the effect that a tool which only support the "min" level can only
+support specific versions of the capabilities it implements. But it also means
+that consumers will need to adjust their behavior to the tool instead of being
+able to ask the tool to adjust to the consumer. Consequently the consumer will
+likely have the more complex logic to do that adjustment.
+
+=== Capabilities
+
+For this proposal capabilities refers to any published coherent target tool
+interface. This can include any single interface, like a single target tool
+option. Or it can include a collective interface of the target tool that covers
+many options. A capability is specified as a series of "scoped" identifiers
+separated by underscore ("_"). The capability must match this regular expression:
+footnote:Regex[ECMAScript® 2022 language specification, 13th edition, June 2022 (https://www.ecma-international.org/publications-and-standards/standards/ecma-262/)]
+
+[source,plan_text]
+----
+^[a-z0-9]+(_[a-z0-9]+)+$
+----
+
+At minimum a capability has two components. The first component is a general
+scope that identifies if the capability is one in the standard, or if it's a
+tool vendor capability.
+
+Standard:: A capability with a scope of `std` indicates that it's defined in the
+EcoStd. footnote:EcoStd[Ecosystem Standards (https://ecostd.github.io/)]
+
+Vendor:: Any other capability, i.e. other than `std`, is available for vendors
+to use as extensions outside the EcoStd. footnote:EcoStd[]
+
+There was a question on "Why not alow 0-9 in the name?". Considering this brings
+up the question as to the utility of having numbers in the name. An
+obvious use case is to add versioning to the name, for example `std2`. That is a
+case we want to avoid. As it avoids using the version numbers themselves which
+subverts the spirit of the introspection. Another use case is to cover vendor
+specific names for tools that use a number in their names, for example `b2`,
+`build2`. Because that is a currently existing use case, and that forcing such
+applications that want custom capabilities to create alternate names has
+various drawbacks, yes, we should accept numbers in the names.
+
+=== Version Specification
```

Comment:
The version specification does not take so-called "epochs" into account. These are specified in tools such as RPM and apt-get, and in packaging specifications such as PEP440.